File: tools/todo_dependency_graph.js
#!/usr/bin/env node
'use strict';
/**
 * todo_dependency_graph.js
 *
 * Node.js (pure stdlib, no npm dependencies) port of
 * tools/todo_dependency_graph.py.
 *
 * Parses a TODO.md following the Feature/Task/Subtask ID scheme:
 *
 *   ## Feature: XXXX — <name>
 *   - [ ] **XXXX-YY** ... PREREQ: A(-B)?:C(-D)?[, A(-B)?:C(-D)?...] ... text
 *   - [x] **XXXX-YY.ZZ** ... (subtask, done)
 *
 * and renders a Graphviz dependency graph where each Feature is drawn as a
 * cluster ("box") containing its open (non-"[x]") Tasks/Subtasks as nodes.
 * Finished tasks and fully-finished features are excluded by default.
 *
 * No third-party npm packages are required. The script emits raw DOT text
 * itself and — unless --dot-text-only is given — shells out to the Graphviz
 * `dot` binary to render PNG and/or SVG output directly (SVG needs no
 * external viewer and can be embedded inline in HTML).
 *
 * Edge color/style convention (same as the Python version):
 *   black        same-feature dependency        (XXXX-YY -> XXXX-ZZ)
 *   blue         cross-feature, task-level       (XXXX-YY -> AAAA-BB)
 *   crimson      cross-feature, feature-level    (XXXX-YY -> AAAA, or XXXX -> AAAA)
 *   dashed gray  prerequisite marked "(soft ...)" in the task text
 *
 * Usage:
 *   node todo_dependency_graph.js [--input TODO.md] [--output graph]
 *                                 [--format svg|png|both] [--include-done]
 *                                 [--dot-binary /path/to/dot] [--dot-text-only]
 *
 * Exit codes:
 *   0  success
 *   1  input file not found / unreadable
 *   2  no features or no tasks discovered (nothing to draw)
 *   3  Graphviz `dot` binary not found
 *   4  Graphviz `dot` binary exited with a non-zero status
 */

const fs = require('fs');
const path = require('path');
const { spawnSync } = require('child_process');

const FEATURE_RE = /^##\s*Feature:\s*(\d{4})\s*(?:—|--|-)?\s*(.*)$/;
const TASK_RE = /^-\s*\[([ xup?])\]\s*(?:\*\*)?(\d+-\d+(?:\.\d+)?)(?:\*\*)?\s*(.*)$/;
const PREREQ_BLOCK_RE = /PREREQ:\s*(.+?)(?:\s*(?:—|--)\s|\s*$)/;
const PREREQ_ITEM_RE = /(\d{4}(?:-\d+(?:\.\d+)?)?)\s*:\s*(\d{4}(?:-\d+(?:\.\d+)?)?)/g;
const SOFT_RE = /\(soft\b/i;

const FEATURE_COLORS = [
  '#cfe8ff', '#d6ecff', '#d9f2d9', '#ffe0cc',
  '#ffd6d6', '#ffe680', '#ffe9cc', '#e6dcff',
  '#c9f7f5', '#f7d9e3', '#e3f7c9', '#d9d9f7',
];

function dotQuote(s) {
  return '"' + String(s).replace(/\\/g, '\\\\').replace(/"/g, '\\"') + '"';
}

class Task {
  constructor(id, featureId, mark, text) {
    this.id = id;
    this.featureId = featureId;
    this.mark = mark;
    this.text = text;
    this.prereqs = []; // [ [targetId, isSoft], ... ]
  }
  get done() {
    return this.mark === 'x';
  }
}

class Feature {
  constructor(id, name) {
    this.id = id;
    this.name = name;
    this.tasks = [];
  }
  get done() {
    return this.tasks.length > 0 && this.tasks.every((t) => t.done);
  }
}

/**
 * Parse TODO.md content into an ordered array of Feature instances.
 * @param {string} text raw file content
 * @returns {Feature[]}
 */
function parseTodo(text) {
  const features = new Map();
  const order = [];
  let currentFeature = null;

  const lines = text.split(/\r?\n/);
  for (const rawLine of lines) {
    const line = rawLine.trim();

    const mFeat = FEATURE_RE.exec(line);
    if (mFeat) {
      const fid = mFeat[1];
      const name = (mFeat[2] || '').trim();
      if (!features.has(fid)) {
        features.set(fid, new Feature(fid, name));
        order.push(fid);
      }
      currentFeature = features.get(fid);
      continue;
    }

    const mTask = TASK_RE.exec(line);
    if (mTask && currentFeature) {
      const tid = mTask[1] === undefined ? null : mTask[2]; // guard, unused
      const mark = mTask[1];
      const id = mTask[2];
      const restText = mTask[3] || '';

      const tidFeature = id.split('-', 1)[0];
      const feature = features.get(tidFeature) || currentFeature;

      const task = new Task(id, feature.id, mark, restText);

      const mBlock = PREREQ_BLOCK_RE.exec(restText);
      if (mBlock) {
        const body = mBlock[1];
        const isSoft = SOFT_RE.test(restText);
        let mItem;
        PREREQ_ITEM_RE.lastIndex = 0;
        while ((mItem = PREREQ_ITEM_RE.exec(body)) !== null) {
          const dst = mItem[2];
          task.prereqs.push([dst, isSoft]);
        }
      }

      feature.tasks.push(task);
    }
  }

  return order.map((fid) => features.get(fid));
}

/**
 * Classify a dependency edge's DOT styling category.
 * @param {Task} srcTask
 * @param {string} dstId
 * @returns {'explicit_same'|'explicit_cross'|'implicit_cross'|'soft_cross'}
 */
function classifyEdge(srcTask, dstId) {
  const isSoft = srcTask.prereqs.some(([dst, soft]) => dst === dstId && soft);
  const dstIsFeatureOnly = !dstId.includes('-');
  const sameFeature = dstId.split('-', 1)[0] === srcTask.featureId;

  if (isSoft) return 'soft_cross';
  if (sameFeature && !dstIsFeatureOnly) return 'explicit_same';
  if (dstIsFeatureOnly) return 'implicit_cross';
  return 'explicit_cross';
}

/**
 * Build DOT source text for the (filtered) dependency graph.
 * @param {Feature[]} features
 * @param {boolean} includeDone
 * @returns {{dot: string, nodeCount: number, edgeCount: number, liveFeatureCount: number}}
 */
function buildDot(features, includeDone) {
  const liveFeatures = features.filter((f) => includeDone || !f.done);

  const allTaskIds = new Set();
  for (const f of liveFeatures) {
    for (const t of f.tasks) {
      if (includeDone || !t.done) allTaskIds.add(t.id);
    }
  }
  const liveFeatureIds = new Set(liveFeatures.map((f) => f.id));

  const lines = [];
  lines.push('digraph todo_dependency_graph {');
  lines.push('  rankdir=LR;');
  lines.push('  splines=true;');
  lines.push('  overlap=false;');
  lines.push('  bgcolor="white";');
  lines.push('  pad="0.3";');
  lines.push('  nodesep="0.25";');
  lines.push('  ranksep="0.8";');
  lines.push('  node [shape=box, style="rounded,filled", fontname="Helvetica", fontsize="10", margin="0.08,0.05"];');
  lines.push('  edge [fontname="Helvetica", fontsize="9"];');

  const edges = []; // [src, dst, kind]

  liveFeatures.forEach((feature, idx) => {
    const color = FEATURE_COLORS[idx % FEATURE_COLORS.length];
    const openTasks = feature.tasks.filter((t) => includeDone || !t.done);
    if (openTasks.length === 0) return;

    lines.push(`  subgraph cluster_${feature.id} {`);
    const label = feature.name ? `${feature.id} — ${feature.name}` : feature.id;
    lines.push(`    label=${dotQuote(label)};`);
    lines.push('    style="rounded,filled";');
    lines.push('    color="#777777";');
    lines.push(`    fillcolor=${dotQuote(color)};`);
    lines.push('    penwidth="1.2";');
    lines.push('    node [fillcolor="white"];');
    lines.push(`    ${dotQuote(feature.id)} [label=${dotQuote(feature.id)}, shape=tab, fillcolor=${dotQuote(color)}, style="filled,bold"];`);
    for (const t of openTasks) {
      lines.push(`    ${dotQuote(t.id)} [label=${dotQuote(t.id)}];`);
    }
    lines.push('  }');

    for (const t of openTasks) {
      for (const [dstId] of t.prereqs) {
        const dstIsFeatureOnly = !dstId.includes('-');
        const targetAlive = allTaskIds.has(dstId) || (dstIsFeatureOnly && liveFeatureIds.has(dstId));
        if (!targetAlive) continue; // prerequisite already finished/removed -> drop the edge
        edges.push([t.id, dstId, classifyEdge(t, dstId)]);
      }
    }
  });

  const styles = {
    explicit_same: 'color="black", penwidth="1.2"',
    explicit_cross: 'color="#1f4e79", penwidth="1.4"',
    implicit_cross: 'color="crimson", penwidth="1.5"',
    soft_cross: 'color="gray50", penwidth="1.2", style="dashed"',
  };
  for (const [src, dst, kind] of edges) {
    lines.push(`  ${dotQuote(src)} -> ${dotQuote(dst)} [${styles[kind]}];`);
  }

  lines.push('  subgraph cluster_legend {');
  lines.push('    label="Legend";');
  lines.push('    style="rounded,dashed";');
  lines.push('    color="#999999";');
  lines.push('    "L1" [label="same-feature / explicit", fillcolor="white"];');
  lines.push('    "L2" [label="cross-feature / explicit", fillcolor="white"];');
  lines.push('    "L3" [label="cross-feature / implicit", fillcolor="white"];');
  lines.push('    "L4" [label="cross-feature / soft", fillcolor="white"];');
  lines.push('    "L1" -> "L1" [color="black", penwidth="1.2"];');
  lines.push('    "L2" -> "L2" [color="#1f4e79", penwidth="1.4"];');
  lines.push('    "L3" -> "L3" [color="crimson", penwidth="1.5"];');
  lines.push('    "L4" -> "L4" [color="gray50", penwidth="1.2", style="dashed"];');
  lines.push('  }');
  lines.push('}');

  return {
    dot: lines.join('\n') + '\n',
    nodeCount: allTaskIds.size,
    edgeCount: edges.length,
    liveFeatureCount: liveFeatures.length,
  };
}

/**
 * Render DOT source to a target format by invoking the Graphviz `dot` binary.
 * @param {string} dotBinary path to dot executable
 * @param {string} dotText DOT source
 * @param {string} format 'svg' | 'png' | any graphviz -T format
 * @param {string} outPath destination file path
 */
function renderWithDot(dotBinary, dotText, format, outPath) {
  const result = spawnSync(dotBinary, [`-T${format}`, '-o', outPath], {
    input: dotText,
    encoding: 'utf-8',
  });
  if (result.error) {
    if (result.error.code === 'ENOENT') {
      throw Object.assign(new Error(`Graphviz binary not found: ${dotBinary}`), { code: 'ENOBIN' });
    }
    throw result.error;
  }
  if (result.status !== 0) {
    throw Object.assign(
      new Error(`Graphviz failed (exit ${result.status}) rendering -T${format}: ${result.stderr}`),
      { code: 'EDOTFAIL' }
    );
  }
}

function parseArgs(argv) {
  const args = {
    input: 'TODO.md',
    output: 'todo_dependency_graph',
    format: 'svg',
    includeDone: false,
    dotBinary: 'dot',
    dotTextOnly: false,
  };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    switch (a) {
      case '--input': args.input = argv[++i]; break;
      case '--output': args.output = argv[++i]; break;
      case '--format': args.format = argv[++i]; break;
      case '--include-done': args.includeDone = true; break;
      case '--dot-binary': args.dotBinary = argv[++i]; break;
      case '--dot-text-only': args.dotTextOnly = true; break;
      default:
        process.stderr.write(`WARNING: unrecognized argument: ${a}\n`);
    }
  }
  return args;
}

function main() {
  const args = parseArgs(process.argv.slice(2));

  const inPath = path.resolve(args.input);
  if (!fs.existsSync(inPath) || !fs.statSync(inPath).isFile()) {
    process.stderr.write(`ERROR: input file not found: ${inPath}\n`);
    return 1;
  }

  const text = fs.readFileSync(inPath, 'utf-8');
  const features = parseTodo(text);
  if (features.length === 0) {
    process.stderr.write("ERROR: no '## Feature: XXXX — ...' headings found in input.\n");
    return 2;
  }

  const totalTasks = features.reduce((sum, f) => sum + f.tasks.length, 0);
  if (totalTasks === 0) {
    process.stderr.write("ERROR: features found, but no '- [ ] **XXXX-YY** ...' tasks parsed.\n");
    return 2;
  }

  const { dot, nodeCount, edgeCount, liveFeatureCount } = buildDot(features, args.includeDone);

  const outBase = args.output;
  const dotPath = `${outBase}.dot`;
  fs.writeFileSync(dotPath, dot, 'utf-8');

  const nFeaturesTotal = features.length;
  const nFeaturesDone = features.filter((f) => f.done).length;
  process.stdout.write(
    `Parsed ${nFeaturesTotal} feature(s) (${nFeaturesDone} fully done), ${totalTasks} task(s) total.\n`
  );
  process.stdout.write(
    `Built graph with ${nodeCount} open task node(s), ${liveFeatureCount} live feature cluster(s), and ${edgeCount} dependency edge(s).\n`
  );
  process.stdout.write(`DOT written to: ${dotPath}\n`);

  if (args.dotTextOnly) {
    return 0;
  }

  const formats = args.format === 'both' ? ['svg', 'png'] : [args.format];
  for (const fmt of formats) {
    const outPath = `${outBase}.${fmt}`;
    try {
      renderWithDot(args.dotBinary, dot, fmt, outPath);
    } catch (err) {
      process.stderr.write(`ERROR: ${err.message}\n`);
      return err.code === 'ENOBIN' ? 3 : 4;
    }
    process.stdout.write(`Output written to: ${outPath}\n`);
  }

  return 0;
}

if (require.main === module) {
  process.exit(main());
}

module.exports = { parseTodo, buildDot, classifyEdge, dotQuote, Task, Feature };

File: tools/todo_dependency_graph.py
#!/usr/bin/env python3
"""
todo_dependency_graph.py

Parses a TODO.md following the Feature/Task/Subtask ID scheme and renders a
Graphviz dependency graph where each Feature is drawn as a cluster ("box")
containing its open (non-"[x]") Tasks/Subtasks as nodes.

No third-party Python packages are required. The script emits raw DOT and then
invokes the Graphviz `dot` binary directly.
"""

import argparse
import re
import shlex
import subprocess
import sys
from pathlib import Path

FEATURE_RE = re.compile(r'^##\s*Feature:\s*(\d{4})\s*(?:\u2014|--|-)?\s*(.*)$')
TASK_RE = re.compile(
    r'^-\s*\[(?P<mark>[ xup?])\]\s*'
    r'(?:\*\*)?(?P<id>\d+-\d+(?:\.\d+)?)(?:\*\*)?\s*'
    r'(?P<text>.*)$'
)
PREREQ_BLOCK_RE = re.compile(r'PREREQ:\s*(?P<body>.+?)(?:\s*(?:\u2014|--)\s|\s*$)')
PREREQ_ITEM_RE = re.compile(r'(\d{4}(?:-\d+(?:\.\d+)?)?)\s*:\s*(\d{4}(?:-\d+(?:\.\d+)?)?)')
SOFT_RE = re.compile(r'\(soft\b', re.IGNORECASE)

FEATURE_COLORS = [
    '#cfe8ff', '#d6ecff', '#d9f2d9', '#ffe0cc',
    '#ffd6d6', '#ffe680', '#ffe9cc', '#e6dcff',
    '#c9f7f5', '#f7d9e3', '#e3f7c9', '#d9d9f7',
]


def dot_quote(s: str) -> str:
    return '"' + s.replace('\\', '\\\\').replace('"', '\\"') + '"'


class Task:
    __slots__ = ('id', 'feature_id', 'mark', 'text', 'prereqs')

    def __init__(self, id_, feature_id, mark, text):
        self.id = id_
        self.feature_id = feature_id
        self.mark = mark
        self.text = text
        self.prereqs = []

    @property
    def done(self):
        return self.mark == 'x'


class Feature:
    __slots__ = ('id', 'name', 'tasks')

    def __init__(self, id_, name):
        self.id = id_
        self.name = name
        self.tasks = []

    @property
    def done(self):
        return len(self.tasks) > 0 and all(t.done for t in self.tasks)


def parse_todo(path: Path):
    features = {}
    order = []
    current_feature = None

    with path.open(encoding='utf-8') as fh:
        for raw_line in fh:
            line = raw_line.rstrip('\n')

            m_feat = FEATURE_RE.match(line.strip())
            if m_feat:
                fid, name = m_feat.group(1), m_feat.group(2).strip()
                if fid not in features:
                    features[fid] = Feature(fid, name)
                    order.append(fid)
                current_feature = features[fid]
                continue

            m_task = TASK_RE.match(line.strip())
            if m_task and current_feature is not None:
                tid = m_task.group('id')
                mark = m_task.group('mark')
                text = m_task.group('text')
                tid_feature = tid.split('-', 1)[0]
                feature = features.get(tid_feature, current_feature)
                task = Task(tid, feature.id, mark, text)

                m_block = PREREQ_BLOCK_RE.search(text)
                if m_block:
                    body = m_block.group('body')
                    is_soft = bool(SOFT_RE.search(text))
                    for m_item in PREREQ_ITEM_RE.finditer(body):
                        _src, dst = m_item.group(1), m_item.group(2)
                        task.prereqs.append((dst, is_soft))

                feature.tasks.append(task)

    return [features[fid] for fid in order]


def classify_edge(src_task: Task, dst_id: str):
    is_soft = any(dst == dst_id and soft for dst, soft in src_task.prereqs)
    dst_is_feature_only = '-' not in dst_id
    same_feature = dst_id.split('-', 1)[0] == src_task.feature_id

    if is_soft:
        return 'soft_cross'
    if same_feature and not dst_is_feature_only:
        return 'explicit_same'
    if dst_is_feature_only:
        return 'implicit_cross'
    return 'explicit_cross'


def build_dot(features, include_done=False):
    live_features = [f for f in features if include_done or not f.done]

    all_task_ids = set()
    for f in live_features:
        for t in f.tasks:
            if include_done or not t.done:
                all_task_ids.add(t.id)
    live_feature_ids = {f.id for f in live_features}

    lines = []
    lines.append('digraph todo_dependency_graph {')
    lines.append('  rankdir=LR;')
    lines.append('  splines=true;')
    lines.append('  overlap=false;')
    lines.append('  bgcolor="white";')
    lines.append('  pad="0.3";')
    lines.append('  nodesep="0.25";')
    lines.append('  ranksep="0.8";')
    lines.append('  node [shape=box, style="rounded,filled", fontname="Helvetica", fontsize="10", margin="0.08,0.05"];')
    lines.append('  edge [fontname="Helvetica", fontsize="9"];')

    edges = []

    for idx, feature in enumerate(live_features):
        color = FEATURE_COLORS[idx % len(FEATURE_COLORS)]
        open_tasks = [t for t in feature.tasks if include_done or not t.done]
        if not open_tasks:
            continue

        lines.append(f'  subgraph cluster_{feature.id} {{')
        label = f'{feature.id} — {feature.name}' if feature.name else feature.id
        lines.append(f'    label={dot_quote(label)};')
        lines.append('    style="rounded,filled";')
        lines.append('    color="#777777";')
        lines.append(f'    fillcolor={dot_quote(color)};')
        lines.append('    penwidth="1.2";')
        lines.append('    node [fillcolor="white"];')
        lines.append(f'    {dot_quote(feature.id)} [label={dot_quote(feature.id)}, shape=tab, fillcolor={dot_quote(color)}, style="filled,bold"];')
        for t in open_tasks:
            lines.append(f'    {dot_quote(t.id)} [label={dot_quote(t.id)}];')
        lines.append('  }')

        for t in open_tasks:
            for dst_id, _is_soft in t.prereqs:
                dst_is_feature_only = '-' not in dst_id
                target_alive = dst_id in all_task_ids or (dst_is_feature_only and dst_id in live_feature_ids)
                if not target_alive:
                    continue
                edges.append((t.id, dst_id, classify_edge(t, dst_id)))

    styles = {
        'explicit_same':  'color="black", penwidth="1.2"',
        'explicit_cross': 'color="#1f4e79", penwidth="1.4"',
        'implicit_cross': 'color="crimson", penwidth="1.5"',
        'soft_cross':     'color="gray50", penwidth="1.2", style="dashed"',
    }
    for src, dst, kind in edges:
        lines.append(f'  {dot_quote(src)} -> {dot_quote(dst)} [{styles[kind]}];')

    lines.append('  subgraph cluster_legend {')
    lines.append('    label="Legend";')
    lines.append('    style="rounded,dashed";')
    lines.append('    color="#999999";')
    lines.append('    "L1" [label="same-feature / explicit", fillcolor="white"];')
    lines.append('    "L2" [label="cross-feature / explicit", fillcolor="white"];')
    lines.append('    "L3" [label="cross-feature / implicit", fillcolor="white"];')
    lines.append('    "L4" [label="cross-feature / soft", fillcolor="white"];')
    lines.append('    "L1" -> "L1" [color="black", penwidth="1.2"];')
    lines.append('    "L2" -> "L2" [color="#1f4e79", penwidth="1.4"];')
    lines.append('    "L3" -> "L3" [color="crimson", penwidth="1.5"];')
    lines.append('    "L4" -> "L4" [color="gray50", penwidth="1.2", style="dashed"];')
    lines.append('  }')
    lines.append('}')

    node_count = len(all_task_ids)
    edge_count = len(edges)
    return '\n'.join(lines) + '\n', node_count, edge_count, len(live_features)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--input', default='TODO.md')
    ap.add_argument('--output', default='todo_dependency_graph')
    ap.add_argument('--format', default='png')
    ap.add_argument('--include-done', action='store_true')
    ap.add_argument('--dot-binary', default='dot')
    args = ap.parse_args()

    in_path = Path(args.input)
    if not in_path.is_file():
        print(f'ERROR: input file not found: {in_path}', file=sys.stderr)
        return 1

    features = parse_todo(in_path)
    if not features:
        print("ERROR: no '## Feature: XXXX — ...' headings found in input.", file=sys.stderr)
        return 2

    total_tasks = sum(len(f.tasks) for f in features)
    if total_tasks == 0:
        print("ERROR: features found, but no '- [ ] **XXXX-YY** ...' tasks parsed.", file=sys.stderr)
        return 2

    dot_text, n_nodes, n_edges, n_live_features = build_dot(features, include_done=args.include_done)
    out_base = Path(args.output)
    dot_path = out_base.with_suffix('.dot')
    out_path = out_base.with_suffix(f'.{args.format}')
    dot_path.write_text(dot_text, encoding='utf-8')

    cmd = [args.dot_binary, f'-T{args.format}', str(dot_path), '-o', str(out_path)]
    try:
        subprocess.run(cmd, check=True)
    except FileNotFoundError:
        print(f"ERROR: Graphviz binary not found: {args.dot_binary}", file=sys.stderr)
        return 3
    except subprocess.CalledProcessError as e:
        print(f"ERROR: Graphviz failed with exit code {e.returncode}: {' '.join(shlex.quote(x) for x in cmd)}", file=sys.stderr)
        return 4

    n_features_total = len(features)
    n_features_done = sum(1 for f in features if f.done)
    print(f"Parsed {n_features_total} feature(s) ({n_features_done} fully done), {total_tasks} task(s) total.")
    print(f"Rendered graph with {n_nodes} open task node(s), {n_live_features} live feature cluster(s), and {n_edges} dependency edge(s).")
    print(f"DOT written to: {dot_path}")
    print(f"Output written to: {out_path}")
    return 0


if __name__ == '__main__':
    sys.exit(main())
